### PR TITLE
Reduce heartbeatbuild logic duplication

### PR DESCRIPTION
**Pull request by onsriahi14**

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/Server.scala [hunk 1]
```diff
@@ -45,12 +45,13 @@ object Server {
...
       val pubSubMediatorRef: ActorRef = system.actorOf(PubSubMediator.props, "PubSubMediator")
 
-      val dbActorRef: AskableActorRef = system.actorOf(Props(DbActor(pubSubMediatorRef, messageRegistry)), "DbActor")
+      val dbActorRef: ActorRef = system.actorOf(Props(DbActor(pubSubMediatorRef, messageRegistry)), "DbActor")
       val securityModuleActorRef: AskableActorRef = system.actorOf(Props(SecurityModuleActor(RuntimeEnvironment.securityPath)))
 
       // Create necessary actors for server-server communications
-      val heartbeatGenRef: ActorRef = system.actorOf(HeartbeatGenerator.props(dbActorRef))
-      val monitorRef: ActorRef = system.actorOf(Monitor.props(heartbeatGenRef))
+      // val heartbeatGenRef: ActorRef = system.actorOf(HeartbeatGenerator.props(dbActorRef))
```
Comment (t1b00) [inline be2-scala/src/main/scala/ch/epfl/pop/Server.scala:52]: Could be good to remove the unnecessary code :)

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 2]
```diff
@@ -51,7 +58,16 @@ final case class Monitor(
     case Monitor.TriggerHeartbeat =>
       log.info("triggering a heartbeat")
       timers.cancel(singleHbKey)
-      heartbeatGenRef ! Monitor.GenerateAndSendHeartbeat(connectionMediatorRef)
+
+      val askForHeartbeat = dbActorRef ? DbActor.GenerateHeartbeat()
+      val heartbeat: HashMap[Channel, Set[Hash]] =
+        Await.ready(askForHeartbeat, duration).value.get match
+          case Success(DbActor.DbActorGenerateHeartbeatAck(Some(map))) => map
+          case Failure(ex: DbActorNAckException)                       => HashMap.empty[Channel, Set[Hash]] // Specific failure
```
Comment (t1b00) [inline be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:66]: Why keep the specific failure if the handling is the same as the case _? Isn't it possible to combine both?

File: be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala [hunk 3]
```diff
@@ -369,6 +370,21 @@ final case class DbActor(
     (publicKey, privateKey)
   }
 
+  @throws[DbActorNAckException]
+  private def generateHeartbeat(): Option[HashMap[Channel, Set[Hash]]] = {
+    val setOfChannels = getAllChannels
+    if (setOfChannels.isEmpty) return None
+    var heartbeatMap: HashMap[Channel, Set[Hash]] = HashMap()
+    setOfChannels.foreach(channel => {
```
Comment (t1b00) [inline be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala:378]: I think you can replace this by a setOfChannels.map and avoid the HashMap creation

**Review by t1b00 — COMMENTED** [review]
There's a few minor code quality tweaks that are possible but other than that LGTM!

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala
Comment (K1li4nL) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:1]: The following test are causing some `Future timed out after [1 second]` logs which are not very reassuring: 

- [ ] "monitor should not schedule any heartbeats when receiving a heartbeat or get_messages_by_id"
- [ ] "monitor should send heartbeats if and only if servers are connected"
- [ ] "monitor should send ConnectTo() requests to ConnectionMediator upon relevant config file change besides first read"

**Review by K1li4nL — CHANGES_REQUESTED** [review]
Lots of moving parts in the code you are touching, good job overall ! Some nitpicks and some simplification that I believe could make the code better

**Review by t1b00 — APPROVED** [review]

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala [hunk 4]
```diff
@@ -24,41 +24,29 @@ object ParamsWithMapHandler extends AskPatternConstants {
...
-        }
-      })
+      val ask = dbActorRef ? DbActor.GenerateHeartbeat()
+      Await.ready(ask, duration).value.get match
+        case Success(DbActor.DbActorGenerateHeartbeatAck(map)) =>
+          map.foreach { m =>
+            m.foreach { case (channel, hashes) =>
+              localHeartBeat(channel) = hashes
+            }
+          }
```
[the remark this answers is not in the document]
Comment (onsriahi14) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala:35]: map is immutable and localHeartbeat is mutable so i cannot assign map directly to localHeartbeat that's why i copied it

[the remark this answers is not in the document]
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala:35]: From what I see in the rest of the code, we don't need it to be mutable ?

File: be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala [hunk 5]
```diff
@@ -369,6 +370,21 @@ final case class DbActor(
     (publicKey, privateKey)
   }
 
+  @throws[DbActorNAckException]
+  private def generateHeartbeat(): Option[HashMap[Channel, Set[Hash]]] = {
```
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala:374]: I feel like an `Option` is not necessary here, you might as well return an empty map directly ?

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 6]
```diff
@@ -170,5 +224,49 @@ class MonitorSuite extends TestKit(ActorSystem("MonitorSuiteActorSystem")) with
...
+    // Ping monitor to inform it of ConnectionMediatorRef
+    mockConnectionMediator.send(monitorRef, ConnectionMediator.Ping())
+    mockConnectionMediator.expectMsgType[ConnectionMediator.ConnectTo](timeout)
+
+    mockConnectionMediator.send(monitorRef, Monitor.TriggerHeartbeat)
+    mockConnectionMediator.expectMsg(timeout, Heartbeat(expected))
+
+    monitorRef ! PoisonPill
+    mockConnectionMediator.expectTerminated(monitorRef)
+
```
Comment (K1li4nL) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:250]: blank line

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 7]
```diff
@@ -170,5 +224,49 @@ class MonitorSuite extends TestKit(ActorSystem("MonitorSuiteActorSystem")) with
...
+    mockConnectionMediator.send(monitorRef, ConnectionMediator.Ping())
+    mockConnectionMediator.expectMsgType[ConnectionMediator.ConnectTo](timeout)
+
+    mockConnectionMediator.send(monitorRef, Monitor.TriggerHeartbeat)
+    mockConnectionMediator.expectNoMessage(timeout)
+
+    monitorRef ! PoisonPill
+    mockConnectionMediator.expectTerminated(monitorRef)
+
```
Comment (K1li4nL) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:269]: blank line

File: be2-scala/src/test/scala/ch/epfl/pop/storage/DbActorSuite.scala [hunk 8]
```diff
@@ -873,4 +877,45 @@ class DbActorSuite extends TestKit(ActorSystem("DbActorSuiteActorSystem")) with
...
+  test("GenerateHearbeat returns an empty localHeartbeat") {
+    val initialStorage: InMemoryStorage = InMemoryStorage()
+    val dbActor: ActorRef = system.actorOf(Props(DbActor(mediatorRef, MessageRegistry(), initialStorage)))
+
+    val ask = dbActor ? GenerateHeartbeat()
+    val answer = Await.result(ask, duration)
+    val heartbeat = answer.asInstanceOf[DbActorGenerateHeartbeatAck].heartbeatMap
+    val expected = HashMap.empty[Channel, Set[Hash]]
+    heartbeat should equal(expected)
+
```
Comment (K1li4nL) [inline be2-scala/src/test/scala/ch/epfl/pop/storage/DbActorSuite.scala:918]: blank line

**Review by K1li4nL — CHANGES_REQUESTED** [review]
Last comments, after that we should be good !

**Review by K1li4nL — APPROVED** [review]
LGTM !

**Pull request merged**